python/rikai/video/ffmpeg.py

```diff
@@ -0,0 +1,116 @@
+#  Copyright 2021 Rikai Authors
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+"""FFmpeg related types and utils"""
+
+import sys
+import numpy as np
+
+try:
+    import ffmpeg
+except ImportError:
+    raise ValueError(
+        "Couldn't import ffmpeg. Please make sure to "
+        "`pip install ffmpeg-python` explicitly or install "
+        "the correct extras like `pip install rikai[all]`"
+    )
+from rikai.types.vision import Image
+from rikai.types.video import VideoSampler
+
+
+class VideoFrameSampler(VideoSampler):
+    """
+    An Image sampler returning one from every `sample_rate` video frames.
+    """
+
+    def __init__(
+        self,
+        stream: VideoStream,
+        sample_rate: int = 1,
+        start_frame: int = 0,
+        max_samples: int = -1,
+        scale_width: int = -1,
+        scale_height: int = -1,
+    ):
+        """
+        Parameters
+        ----------
+        sample_rate: int
+            The sampling rate in number of frames
+        start_frame: int
+            Start from a specific frame (0-based indexing)
+        max_samples: int
+            Yield at most this many frames (-1 means no max)
+        """
+        super().__init__(stream)
+        self.sample_rate = sample_rate
+        self.start_frame = start_frame
+        self.max_samples = max_samples
+        self.scale_width = (scale_width,)
+        self.scale_height = (scale_height,)
+        self.probe = None
+
+    def _probe(self):
+        try:
+            self.probe = ffmpeg.probe(self.stream.uri)
+        except ffmpeg.Error as e:
+            print(e.stderr, file=sys.stderr)
```

Comment:
you'll want to use a logger instead of bare printing (see logger usage in other parts of the code base for examples). Log at error level. Also, I don't know if you'll want to log the entire ffmpeg output -- is there a way to just get a message + stacktrace?

Reply:
Found something related while looking into SSIM:
![image](https://user-images.githubusercontent.com/9044907/113436644-f2570280-9399-11eb-9c47-31615014e2c6.png)